camera exposure time in lightsheet mode

annie-xd-wang opened the pull request:
#630: The error has been fixed for the Hamamatsu Orca Fire camera. It was due to inaccuracies in exposure time calculation for that specific camera. If the same error occurs on other machines, I'll need to read the specific documentation.

AdvancedImagingUTSW committed:
camera exposure time in lightsheet mode

Co-Authored-By: Annie Wang <6161065+annie-xd-wang@users.noreply.github.com>

zacsimile commented:
Awesome! That makes sense. Specs for 1H can be found for the Flash (our current assumed default) at https://www.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C13440-20CU_IM_En.pdf, the  Fusion BT at https://www.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C15440-20UP,-20UP01_IM_En.pdf and the Lightning at https://www.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C14120-20P_IM_En.pdf. 

It's worth noting that 1H changes depending on the camera's mode of operation and whether or not is is camera link vs. USB. My original thought was we could just let it throw the error, since I believe this is from our side of the code (which has a fixed rounding tolerate =/= 1H, line 1043 of `HamamatsuAPI.py`), since the camera automatically then does its rounding, and we pull the true exposure time for whatever the current 1H is back from the camera itself. Effectively, if we divide by a particular 1H for whatever condition we most commonly use, it will work for that condition and then still throw this error if the camera's mode of operation is changed.

zacsimile commented:
Looking at the code more closely, I don't fully understand why this fixes the error. Is `properly_value_step` as defined in line 1007 of `HamamatsuAPI.py` not equivalent to 1H? I think it should be. Perhaps exposure time needs to be set before line interval...

AdvancedImagingUTSW committed:
Update camera_hamamatsu.py

@annie-xd-wang, @Sriya-235 and I were going over some of the basics of code review, and I probably went a bit too crazy.

So, each type of Hamamatsu Camera has its own 1H/minimum exposure time value. Thus, it seems as if we should have this specified in their very own __init__ functions.

Then we were a bit confused as well. Why is it that the methods for max_image_width, which are getters and setters, are both in the parent and child classes? It seems like we could eliminate a fair amount of code if these were only defined in the parent class.  We created a new HamamatsuBase class in hopes of simplifying the children classes...  For example, I think the calculate_exposure_time is only unique for the Hamamatsu Lightning camera...

annie-xd-wang committed:
refactor

annie-xd-wang commented:
@zacsimile, I agree with you that properly_value_step, defined in line 1007 of the HamamatsuAPI, is intended to round up 1H. However, it didn't behave that way the last time I worked with the camera. I will double-check the value of properly_value_step for exposure time the next time I connect to a camera.

annie-xd-wang committed:
Merge branch 'develop' into fix_exposure_time_error

annie-xd-wang committed:
keep x_pixels and y_pixels

Sriya-235 committed:
exposure time in light-sheet mode

Co-Authored-By: Annie Wang <6161065+annie-xd-wang@users.noreply.github.com>

Sriya-235 committed:
small fix

Co-Authored-By: Annie Wang <6161065+annie-xd-wang@users.noreply.github.com>

annie-xd-wang committed:
update the line interval calculation for Hamamatsu Fire

annie-xd-wang commented:
The minimum exposure time we get from the camera is 1H. Casting the exposure time according to the step size causes accuracy errors.

Sriya-235 committed:
Quick updates

Evaluating new light-sheet calculation for ORCA FIRE

zacsimile commented:
Nice find! Super weird that rounding to integer multiples the step size doesn't work, but glad it's working with the minimum now.

annie-xd-wang committed:
using fixed time

annie-xd-wang committed:
use milliseconds as the delay unit

annie-xd-wang committed:
calculate framerate in controller side

annie-xd-wang committed:
remove get_readout_function()

annie-xd-wang committed:
rename calculate_readout_time in controllers

annie-xd-wang committed:
update readout time and line interval calculation for Hamamatsu Fire

annie-xd-wang committed:
using seconds in camera functions

annie-xd-wang committed:
update exposure time in light-sheet mode

annie-xd-wang committed:
Merge branch 'develop' into fixed-timing

annie-xd-wang committed:
automatically use minimum camera delay

annie-xd-wang committed:
verify configuration

annie-xd-wang committed:
Revert "automatically use minimum camera delay"

This reverts commit 387e583a37ffa0941e6ba44f82d19f26c596c88c.

annie-xd-wang committed:
add fly back time in waveform popup window

annie-xd-wang committed:
round up displaying exposure time

annie-xd-wang committed:
update tests

annie-xd-wang committed:
Update test_model.py

annie-xd-wang committed:
Merge branch 'develop' into fix_exposure_time_error

annie-xd-wang committed:
update tests

annie-xd-wang committed:
update tests

annie-xd-wang committed:
update tests

annie-xd-wang committed:
fix bug: serial number with leading zero but not an octal number

annie-xd-wang committed:
update line interval calculation for Hamamatsu Lightning and Fusion

annie-xd-wang committed:
galvo frequency

annie-xd-wang committed:
Merge branch 'camera_serial_number_with_leading_zeros' into fix_exposure_time_error

AdvancedImagingUTSW commented:
I quickly tried to test it on a Mac, and I got the following error when running it in the continuous mode. However, when I tried to get the same error again, it did not happen. Seems somewhat random. Was in continuous mode, 3 channels selected, hit stop, and it threw the error.

```
Exception in thread live signal:
Traceback (most recent call last):
  File "/opt/miniconda3/envs/navigate/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/opt/miniconda3/envs/navigate/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/S155475/Desktop/GitHub/navigate/src/navigate/model/model.py", line 1094, in run_live_acquisition
    self.signal_container.reset()
AttributeError: 'Model' object has no attribute 'signal_container'

```

AdvancedImagingUTSW commented on src/navigate/controller/sub_controllers/camera_setting_controller.py:
```diff
@@ -464,43 +463,23 @@ def calculate_physical_dimensions(self):
         elif sensor_mode == "Light-Sheet":
             #  Progressive sensor mode operation
-            readout_time = exposure_time - 1 / (exposure_time + (vn + 10) * h)
+            readout_time = 0
```
Is this always the case? What happens if the exposure time is very short (e.g., 1 ms)?

AdvancedImagingUTSW commented on src/navigate/model/device_startup_functions.py:
```diff
@@ -1240,7 +1240,11 @@ def load_devices(configuration, is_synthetic=False, plugin_devices={}) -> dict:
-                    devices["camera"][build_ref_name("_", device["type"], int(camera_serial_number, 8))] = camera
+                    try:
+                        oct_num = int(camera_serial_number, 8)
+                        devices["camera"][build_ref_name("_", device["type"], oct_num)] = camera
```
I still need to test this.
